Automattic/jetpack
# Fix: protect wpcom urls from search and replace

IDC logic stores wpcom URLs to the database and uses them to display Safe Mode. Users sometimes use search & replace plugins to alter URLs on cloned sites and rewrite these values which causes domain change on wpcom side.

## Proposed changes:

* Using strrev() to reverse stored domains but keep them readable for debugging purposes before storing sync_error_idc option.
* Add a method to reverse options for comparison.
* Used `jetpack-options` filter to apply reversing and re-reversing
* Updated tests

### Other information:


## Jetpack product discussion


pf5801-3R-p2

## Does this pull request change what data or activity we track or use?



## Testing instructions:






* You will need two JN sites with connected Jetpack.
* Use Jetpack beta to load this branch.
* Clone database from site A and import it to the site B. 
*  It should trigger Safe Mode. If not, try opening the Jetpack dashboard or My Jetpack.
* Ignore Safe Mode and install a search and replace plugin such as[ Better search replace](https://wordpress.org/plugins/better-search-replace/). Use the plugin to update all mentions of URL A with URL B in wp-options table.
* Safe Mode should still show up.
* Check audit logs and debug to make sure URLs didn't change.

## Alternative testing

* Create JN site, connect it and run this branch
* Run IDC tool from debugger tools
* Once you trigger Safe Mode, check the database for jetpack_sync_error_idc option. It should contain reversed wpcom values.
